Exprimental WASI support for ports/unix

yamt opened the pull request:
~~See comments in build-wasi.sh~~

yamt committed:
unix: Add wasi variant.

Signed-off-by: YAMAMOTO Takashi <yamamoto@midokura.com>

yamt committed:
unix/unix_mphal: Disable signal mask stuff for WASI.

WASI dosen't have signals.

Signed-off-by: YAMAMOTO Takashi <yamamoto@midokura.com>

stinos commented on ports/unix/mpconfigport.h:
```diff
@@ -154,8 +154,12 @@ typedef long mp_off_t;
 // Don't default sys.argv and sys.path because we do that in main.
 #define MICROPY_PY_SYS_PATH_ARGV_DEFAULTS (0)
 
+#if defined(__wasi__)
```
Should be done like 

```
#ifndef  MICROPY_PY_SYS_EXECUTABLE
#define MICROPY_PY_SYS_EXECUTABLE (1)
#endif
```

so it works generically, then make a separate mpconfigvariant_wasi.h where it gets set to t 0.

yamt replied:
who would include mpconfigvariant_wasi.h?

stinos replied:
That would be done automatically if you use `make VARIANT=wasi` (though, my mistake, it should actually be unix/variants/wasi/mpconfigvariant.h). Likewise instead of build.sh you'd implement the specific makefile options in unix/variants/wasi/mpconfigvariant.mk). Check out the other variants for examples. Also I think to have any chance of merging this ulab should probably not be included unconditionally because this makes it harder to build out of the box; on the other hand if this only works with non-approved upstream wasi patches it's not super useful for others anyway.

dpgeorge replied:
Yes, please make this a new variant.  It should then be possible to build it via `make VARIANT=wasi`.

yamt replied:
ok. i was not aware of the mechanism. thank you. let me try to adapt.

yamt replied:
> Also I think to have any chance of merging this ulab should probably not be included unconditionally because this makes it harder to build out of the box; on the other hand if this only works with non-approved upstream wasi patches it's not super useful for others anyway.

sure.

wrt wasi patches, alternatively it's possible to put the setjmp/longjmp runtime within micropython itself. at least until when setjmp/longjmp gets ready in wasi. how do you think?

anyway, it won't be super useful until wasm EH is widely implemented in runtimes though.

dpgeorge replied:
If building unix for wasi requires patches to wasi/binaryen then IMO this PR should stay as a draft until those patches are no longer needed.  You could put the setjmp/longjmp code here, but I guess the other patches are still needed.

yamt replied:
i adapted to use VARIANT. it simplify a lot. thank you.

i'm not sure what should be done for UNAME_S logic in Makefile though.
a build for WASI is actually a cross compile. but the existing CROSS_COMPILE seems geared to "arm-none-eabi-gcc" style cross toolchain, which doesn't seem suitable to wasi-sdk style cross toolchain.

yamt replied:
> If building unix for wasi requires patches to wasi/binaryen then IMO this PR should stay as a draft until those patches are no longer needed. You could put the setjmp/longjmp code here, but I guess the other patches are still needed.

FYI, all toolchain patches have been merged in the upstream since then.

stinos commented on ports/unix/variants/mpconfigvariant_common.h:
```diff
@@ -91,7 +91,11 @@
 #define MICROPY_PY_OS_ERRNO            (1)
 #define MICROPY_PY_OS_GETENV_PUTENV_UNSETENV (1)
 #define MICROPY_PY_OS_SEP              (1)
+#if defined(__wasi__)
```
Similar comment as above: instead of modifying this variant it would be better to create it separately

dpgeorge commented:
See semi-related #11723, which gets `mpy-cross` compiling to WASM (but not WASI at this stage).

yamt committed:
unix: Make a few things overridable.

Make the following items overridable:

    MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
    MICROPY_PY_OS_SYSTEM
    MICROPY_PY_SYS_EXECUTABLE
    MICROPY_PY_WEBSOCKET

I plan to override them for wasi.

Signed-off-by: YAMAMOTO Takashi <yamamoto@midokura.com>

yamt commented:
is there any TODO for this PR while waiting for LLVM 19?

yamt commented:
the last push is a conflict resolution

yamt commented:
now wasi-sdk 25 has been released and this only requires released versions of the toolchain.
can you consider to make this land?

yamt commented:
the ci failure doesn't seem related to this PR.
```
Cloning into '/home/runner/work/micropython/micropython/lib/asf4'...
fatal: unable to access 'https://github.com/adafruit/asf4/': The requested URL returned error: 403
fatal: clone of 'https://github.com/adafruit/asf4' into submodule path '/home/runner/work/micropython/micropython/lib/asf4' failed
Failed to clone 'lib/asf4'. Retry scheduled
```

yamt commented:
@dpgeorge @stinos  does this have any blockers?

stinos commented on ports/unix/variants/wasi/mpconfigvariant.mk:
```diff
@@ -0,0 +1,28 @@
+FROZEN_MANIFEST ?= $(VARIANT_DIR)/manifest.py
+
+# WASI doesn't have FFI
+MICROPY_PY_FFI=0
```
Should have whitespace around the ` = `

stinos commented on ports/unix/variants/wasi/mpconfigvariant.mk:
```diff
@@ -0,0 +1,28 @@
+# Build for WASI
```
Imo not needed to state the obvious.,

stinos commented on ports/unix/variants/wasi/manifest.py:
```diff
@@ -0,0 +1,3 @@
+# include("$(PORT_DIR)/variants/manifest.py")
```
Since this file only has comments can it just be omitted?

stinos commented on ports/unix/unix_mphal.c:
```diff
@@ -43,7 +43,13 @@
+#define HAS_SIGNAL 1
+#endif
+
+#if HAS_SIGNAL != 0
```
Just `#if HAS_SIGNAL` is good enough?

yamt replied:
done

stinos commented on ports/unix/build-wasi.sh:
```diff
@@ -0,0 +1,55 @@
+PROG=build-wasi/micropython
+
+# Unlike emscripten, WASI doesn't provide a way to scan GC roots
+# like WASM locals. Hopefully --spill-pointers can workaround it.
```
Did you try this with e.g. long-running scripts?

yamt replied:
no

stinos commented:
Apart from some cosmetic changes (see somments):
- can you squash this so only relevant commits are left? It probably comes down to a commit which makes some PP defs in the unix port configurable, a commit which fixes some unix build stuff, and then one commit which adds the WASI variant.
- it seems like almost none of the other ports/variants require a .sh file to build them, the goal is to have builds like `make VARIANT=wasi WASI_SDK=/opt/wasi-sdk-25.0`. So I wonder how much of build-wasi.sh is really needed: all build flags can probably go in mpconfigvariant.mk? That would be more consistent with how variants are used in general. And I think, not 100% sure though, you could also get those post-build `${WASM_OPT}` call in there by changing the `STRIP` definition to first strip then call wasm-opt.
- in general ports/variants shouldn't go stale and the first step in making that happen is having the CI build it so errors can be caught early. Which would come down to building the WASI variant as part of the github workflows (see tools/ci.sh and workflows/ports_unix.yml), and then running it, for instance have it run all unit tests. Would that be possible somehow?

yamt commented:
@stinos thank you for comments. i will take my time to digest.

yamt commented:
> Apart from some cosmetic changes (see somments):
> 
>     * can you squash this so only relevant commits are left? It probably comes down to a commit which makes some PP defs in the unix port configurable, a commit which fixes some unix build stuff, and then one commit which adds the WASI variant.

i will consider squashing later.
maybe one of the last things to do.
 
>     * it seems like almost none of the other ports/variants require a .sh file to build them, the goal is to have builds like `make VARIANT=wasi WASI_SDK=/opt/wasi-sdk-25.0`. So I wonder how much of build-wasi.sh is really needed: all build flags can probably go in mpconfigvariant.mk? That would be more consistent with how variants are used in general. And I think, not 100% sure though, you could also get those post-build `${WASM_OPT}` call in there by changing the `STRIP` definition to first strip then call wasm-opt.

partly done.
but i'm not sure if it's a good idea to abuse STRIP for this purpose
because STRIP seems to be documented as user-overridable.

>     * in general ports/variants shouldn't go stale and the first step in making that happen is having the CI build it so errors can be caught early. Which would come down to building the WASI variant as part of the github workflows (see tools/ci.sh and workflows/ports_unix.yml), and then running it, for instance have it run all unit tests. Would that be possible somehow?

ok.

i have run some of test cases in this repo with toywasm locally.

yamt committed:
tests/run-tests.py: Don't assume native execution for wasm.

Signed-off-by: YAMAMOTO Takashi <yamamoto@midokura.com>

yamt committed:
tools/ci.sh: Add wasi case.

Signed-off-by: YAMAMOTO Takashi <yamamoto@midokura.com>

yamt committed:
ports_unix.yml: Add wasi case.

Signed-off-by: YAMAMOTO Takashi <yamamoto@midokura.com>

yamt commented:
the last push was a conflict resolution.

yamt commented:
@dpgeorge @stinos can you revisit when/if you have time? thank you.

yamt committed:
unix: Move some wasi-specific rules to wasi/mpconfigvariant.mk.

Signed-off-by: YAMAMOTO Takashi <yamamoto@midokura.com>

yamt commented:
> > but i thought mpconfigvariant.mk was meant only for declarative things like setting variables
> 
> That's mostly how it gets currently used, but I don't think that it's a fixed rule, could put anything in there

ok. i moved the rules. i had to duplicate `PROG ?= micropython`.

yamt commented:
the last push was a mechanical rebase

yamt commented:
to me, the ci failure (Address already in use / Connection refused) seems unrelated to the changes in this PR.

yamt commented:
the last push was a conflict resolution.

matrey commented:
FYI I just stumbled upon this: https://github.com/agentskillmania/build-micro-python-wasi/blob/main/micro-python-wasi.README.md "MicroPython (v1.27.0) compiled to WebAssembly (WASI Preview2) for sandboxed execution."
